contracts/0.8.9/StakingRouter.sol

```diff
@@ -0,0 +1,523 @@
+     */
+    function setWithdrawalCredentials(bytes32 _withdrawalCredentials)
+        external
+        onlyRole(MANAGE_WITHDRAWAL_CREDENTIALS_ROLE)
+    {
+        WITHDRAWAL_CREDENTIALS_POSITION.setStorageBytes32(_withdrawalCredentials);
+
+        //trim keys with old WC
+        _trimUnusedKeys();
+
+        emit WithdrawalCredentialsSet(_withdrawalCredentials);
+    }
+
+    /**
+     * @notice Returns current credentials to withdraw ETH on ETH 2.0 side after the phase 2 is launched
+     */
+    function getWithdrawalCredentials() public view returns (bytes32) {
+        return WITHDRAWAL_CREDENTIALS_POSITION.getStorageBytes32();
+    }
+
+    function _trimUnusedKeys() internal {
+        uint256 stakingModulesCount = getStakingModulesCount();
+        for (uint256 i = 0; i < stakingModulesCount; ++i) {
+            IStakingModule(_getStakingModuleAddressByIndex(i)).trimUnusedKeys();
+        }
+    }
+
+    function _loadStakingModuleCache(uint24 _stakingModuleId)
+        internal
+        view
+        returns (StakingModuleCache memory stakingModuleCache)
+    {
+        StakingModule storage stakingModuleData = _getStakingModuleById(_stakingModuleId);
+        stakingModuleCache.paused = stakingModuleData.paused;
+        stakingModuleCache.targetShare = stakingModuleData.targetShare;
+
+        IStakingModule stakingModule = IStakingModule(stakingModuleData.stakingModuleAddress);
+        (uint256 totalKeysCount, uint256 usedKeysCount, uint256 stoppedKeysCount) = stakingModule.getSigningKeysStats();
+        stakingModuleCache.totalKeysCount = totalKeysCount;
+        stakingModuleCache.usedKeysCount = usedKeysCount;
+        stakingModuleCache.stoppedKeysCount = stoppedKeysCount;
+        stakingModuleCache.activeKeysCount = stakingModuleCache.usedKeysCount - stakingModuleCache.stoppedKeysCount;
+        stakingModuleCache.availableKeysCount = stakingModuleCache.totalKeysCount - stakingModuleCache.usedKeysCount;
+    }
+
+    function _getKeysAllocation(uint256 _activeKeysCount, uint256 _keysToAllocate)
+        public
+        view
+        returns (uint256 allocated, uint256[] memory allocations)
+    {
+        uint256 stakingModulesCount = getStakingModulesCount();
+        allocations = new uint256[](stakingModulesCount);
+        uint256[] memory capacities = new uint256[](stakingModulesCount);
+        uint256 _totalActiveKeysCount = _activeKeysCount + _keysToAllocate;
+
+        for (uint256 i = 0; i < stakingModulesCount; ++i) {
+            StakingModuleCache memory stakingModuleCache = _loadStakingModuleCache(_getStakingModuleIdByIndex(i));
+            allocations[i] = stakingModuleCache.activeKeysCount;
+
+            uint256 targetKeys = (stakingModuleCache.targetShare * _totalActiveKeysCount) / TOTAL_BASIS_POINTS;
+            capacities[i] = Math.min(
+                targetKeys,
+                stakingModuleCache.activeKeysCount + stakingModuleCache.availableKeysCount
+            );
```

Comment:
Shouldn't we exclude modules that are paused or set their capacity to 0 during allocation?